src/app/rosetta/lib/block.ml

```diff
@@ -0,0 +1,579 @@
+  module Kind = struct
+    type t = [`Coinbase | `Fee_transfer] [@@deriving eq]
+  end
+
+  type t =
+    { kind: Kind.t
+    ; receiver: [`Pk of string]
+    ; fee: Unsigned.UInt64.t
+    ; token: Unsigned.UInt64.t
+    ; hash: string }
+
+  let to_operations ~user_commands ~coinbase ~block_creator_pk (t : t) :
+      Operation.t list =
+    (* We choose to represent the dec-side of fee transfers from txns from the
+     * canonical user command that created them so we are able consistently
+     * produce more balance changing operations in the mempool or a block.
+     * *)
+    let plan : 'a Op.t list =
+      match t.kind with
+      | `Coinbase ->
+          [{Op.label= `Coinbase_inc; related_to= None}]
+      | `Fee_transfer ->
+          (* Detect if a fee transfer comes from a snark worker payment by
+         * failing to find the user command that caused it *)
+          if
+            List.find user_commands ~f:(fun (info : User_command_info.t) ->
+                Unsigned.UInt64.equal info.fee t.fee
+                && Unsigned.UInt64.equal info.token t.token
+                && [%eq: [`Pk of string]] info.receiver t.receiver )
+            |> Option.is_none
+          then
+            (* In this case we make a "synthetic" fee subtraction from block
+           * creator even though this is implicit in our blocks. We need this
+           * so the balance always sums to zero. *)
+            [ {Op.label= `Fee_payer_dec; related_to= None}
+            ; {Op.label= `Fee_receiver_inc; related_to= Some `Fee_payer_dec} ]
+          else
+            (* Otherwise we only have the receiver inc side *)
+            (* TODO: Relate fee_payer_dec with fee_receiver_inc across
+           * user_commands and internal commands *)
+            [{Op.label= `Fee_receiver_inc; related_to= Some `Fee_payer_dec}]
+    in
+    Op.build ~a_eq:[%eq: [`Coinbase_inc | `Fee_payer_dec | `Fee_receiver_inc]]
+      ~plan ~f:(fun ~related_operations ~operation_identifier op ->
+        (* All internal commands succeed if they're in blocks *)
+        let status = Operation_statuses.name `Success in
+        match op.label with
+        | `Coinbase_inc ->
+            { Operation.operation_identifier
+            ; related_operations
+            ; status
+            ; account=
+                Some (account_id block_creator_pk Amount_of.Token_id.default)
+            ; _type= Operation_types.name `Coinbase_inc
+            ; amount= Some (Amount_of.coda coinbase)
+            ; metadata= None }
+        | `Fee_payer_dec ->
+            { Operation.operation_identifier
+            ; related_operations
+            ; status
+            ; account= Some (account_id block_creator_pk t.token)
+            ; _type= Operation_types.name `Fee_payer_dec
+            ; amount= Some Amount_of.(negated @@ token t.token t.fee)
+            ; metadata= None }
+        | `Fee_receiver_inc ->
+            { Operation.operation_identifier
+            ; related_operations
```

Comment:
How do you plan on using the `related_operations` here? It looks like it would have all the `Fee_payer_dec` ops for each `Fee_receiver_inc` op? The relation is `sum(fee-payer-decs) + sum(fee-receiver-incs) = 0`

Reply:
Also I'm going to make the relatedness a TODO for later, I don't think it's important enough to warrant the annoyingness of implementing it now